### PR TITLE
Check stack height for OP_CALLF validation

### PR DESCRIPTION
**Pull request by rodiazet**
- Add validation rule.
 - Add unit tests.
 - Fix already existing unit test.

### CODE REVIEW

File: lib/evmone/eof.cpp [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 #include "eof.hpp"
 #include "baseline_instruction_table.hpp"
+// TODO: Remove this include. It's needed only to use StackSpace::limit
```
Comment (gumb0) [inline lib/evmone/eof.cpp:7]: Probably better to redefine it in this file.

File: test/unittests/evm_eof_function_test.cpp [hunk 2]
```diff
@@ -82,9 +82,12 @@ TEST_P(evm, callf_stack_overflow)
 
     rev = EVMC_CANCUN;
     const auto code =
-        bytecode{"ef0001 010008 020002 0BFF 0007 030000 00 000003FF 00000002"_hex} +  // EOF header
-        1023 * push(1) + OP_CALLF + bytecode{"0x0001"_hex} + 1021 * OP_POP + OP_RETURN +
-        2 * push(1) + 2 * OP_POP + OP_RETF;
+        bytecode{
+            "ef0001 01000c 020003 0BFF 0007 0004 030000 00 000003FF 00000001 00000001"_hex} +  // EOF
+                                                                                               // header
```
Comment (gumb0) [inline test/unittests/evm_eof_function_test.cpp:87]: formatting went funky here, maybe better just delete the comment

**Comment by gumb0** [comment]
State tests currently seem to have a bug https://github.com/ethereum/tests/pull/1238#discussion_r1270786226

**Comment by gumb0** [comment]
Similar validation-time check should be added to JUMPF (but JUMPF is not merged yet: https://github.com/ethereum/evmone/pull/644)

File: lib/evmone/eof.cpp [hunk 3]
```diff
@@ -341,18 +342,23 @@ std::variant<EOFValidationError, int32_t> validate_max_stack_height(
...
+        auto stack_height = stack_heights[i];
+        assert(stack_height != LOC_UNVISITED);
+
         if (opcode == OP_CALLF)
         {
             const auto fid = read_uint16_be(&code[i + 1]);
 
             stack_height_required = static_cast<int8_t>(code_types[fid].inputs);
+
+            if (stack_height + code_types[fid].max_stack_height - stack_height_required >
```
Comment (gumb0) [inline lib/evmone/eof.cpp:354]: In case of stack underflow this expression can be negative, I think it's ok, it's signed and won't underflow

**Comment by gumb0** [comment]
Rebased.

File: lib/evmone/eof.cpp (hunk 3 above)
Comment (gumb0) [reply lib/evmone/eof.cpp:354]: I haven't found any tests for stack underflow validation at all. I will add some at least for CALLF.

Comment (gumb0) [reply lib/evmone/eof.cpp:354]: I was wrong it cannot be negative, because callee's `max_stack_height >= inputs`

File: lib/evmone/eof.cpp [hunk 4]
```diff
@@ -341,18 +342,23 @@ std::variant<EOFValidationError, int32_t> validate_max_stack_height(
         auto stack_height_required = instr::traits[opcode].stack_height_required;
         auto stack_height_change = instr::traits[opcode].stack_height_change;
 
+        auto stack_height = stack_heights[i];
```
Comment (chfast) [inline lib/evmone/eof.cpp:345]:
```suggestion
        const auto stack_height = stack_heights[i];
```

File: test/unittests/eof_validation_test.cpp [hunk 5]
```diff
@@ -733,6 +734,75 @@ TEST(eof_validation, callf_invalid_code_section_index)
         EOFValidationError::invalid_code_section_index);
 }
 
+TEST(eof_validation, callf_stack_overflow)
+{
+    {
+        auto code = eof1_bytecode(
+            512 * push(1) + OP_CALLF + bytecode{"0x0000"_hex} + 510 * OP_POP + OP_RETURN, 512);
```
Comment (chfast) [inline test/unittests/eof_validation_test.cpp:741]:
```suggestion
            512 * push(1) + OP_CALLF + "0000" + 510 * OP_POP + OP_RETURN, 512);
```
`bytecode` can consume strings and will interpret them as hex.

**Review by chfast — APPROVED** [review]

File: lib/evmone/eof.cpp (hunk 4 above)
Comment (rodiazet) [reply lib/evmone/eof.cpp:345]: It's modified in line 365

**Pull request merged**